### PR TITLE
Add e2e test for cooperative refund of underfunded submarine swap

### PR DESCRIPTION
**Pull request by michael1011**
The new test is going to pass after https://github.com/BoltzExchange/boltz-backend/pull/1381 is merged

## Summary by CodeRabbit

* **Tests**
  * Added end-to-end coverage for submarine swap underfunded scenarios, including address/script validation and cooperative refund flow.
* **Refactor**
  * Stabilized e2e tests by replacing fixed sleeps with a retrying settle mechanism and shared helpers for faucet funding and settlement polling, reducing flakiness.
* **Chores**
  * Improved CI diagnostics to collect container/service logs on integration-test failures and added extra debug output for swap failure investigation.

### CODE REVIEW

**Review by ghost — APPROVED** [review]
## Review: Add e2e test for cooperative refund of underfunded submarine swap

**Verdict: APPROVE** — test-only change, no production code modified. Clean, well-structured test that exercises an important edge case.

### What this does

Adds `TestSubmarineSwapUnderfundedCooperativeRefund` which:
1. Creates a submarine swap via Boltz
2. Verifies local VHTLC address derivation matches Boltz's (good correctness check)
3. Deliberately underfunds the swap by 100 sats
4. Waits for Boltz to mark it failed
5. Exercises `RefundSwap` cooperative path (`withReceiver: true`)

Also adds `boltzLocktime` helper for converting Boltz timeout values to `arklib.RelativeLocktime`.

### Findings

**1. `boltzLocktime` threshold (minor)** — `swap_test.go:449-453`

The 512 threshold for seconds-vs-blocks classification is a Boltz convention (BIP68 seconds granularity = 512). It works correctly in practice but is undocumented. A brief comment would help future readers:

```go
// boltzLocktime converts Boltz API timeout values to RelativeLocktime.
// Boltz convention: values >= 512 are seconds (BIP68 granularity), otherwise blocks.
func boltzLocktime(value uint32) arklib.RelativeLocktime {
```

**2. `handlerKey` is distinct from the ark client's key (note, not a bug)** — `swap_test.go:380`

The `handlerKey` is a fresh random key passed to `NewSwapHandler`. This is the handler's signing key for Boltz interactions (stored as `SwapHandler.privateKey`), separate from the ark client's key used for VTXO operations. The separation is correct and matches production flow.

**3. No post-refund balance assertion (nit)** — After the cooperative refund succeeds, the test doesn't verify the funds returned. This is fine for a focused "does the refund succeed" test, but a balance check would strengthen confidence that the full round-trip works. Low priority since timing/settlement makes this flaky.

**4. Preimage hash derivation is correct** — `Ripemd160H(SHA256(preimage))` = HASH160(preimage), matching the VHTLC preimage condition. ✓

**5. External dependency** — PR description notes this depends on [BoltzExchange/boltz-backend#1381](https://github.com/BoltzExchange/boltz-backend/pull/1381). Test will fail in CI until that merges and the test infra picks it up. No action needed, just awareness.

### Cross-repo impact

None. Single test file change. All imported types from `arkd` (`AbsoluteLocktime`, `RelativeLocktime`, `LocktimeTypeSecond`, `LocktimeTypeBlock`, `clientTypes.Receiver`) are stable and verified present in the pinned dependency.

LGTM 🟢

🤖 Reviewed by Arkana

**Review by ghost — APPROVED** [review]
## 🤖 Reviewed by Arkana (incremental — new commit `001f83b4`)

The new "fix test" commit is a clean test infrastructure refactor. Replaces hardcoded `time.Sleep` + settle patterns with retry-based `faucetAndSettle` and `waitForSettle` helpers. No protocol-critical changes.

**Minor nits (non-blocking):**

1. **`utils_test.go:873`** — Missing space before `{`: `amount float64){` → `amount float64) {`. Run `gofmt`.
2. **`utils_test.go:883`** — Unnecessary bare `return` at end of `faucetAndSettle`.
3. **`vhtlc_test.go:841`** — `faucetAndSettle` already calls `strings.TrimSpace(address)` internally, but this call site double-trims with `strings.TrimSpace(boarding[0])`. Other call sites don't. Pick one pattern.
4. **`main_test.go:106-110`** — The closure `func(ctx context.Context) error { _, err := f.Settle(ctx, &pb.SettleRequest{}); if err != nil { return err }; return nil }` simplifies to `_, err := f.Settle(...); return err`.

All test-only, all non-blocking. LGTM.

**Review by ghost — APPROVED** [review]
## 🤖 Reviewed by Arkana (incremental — new commit `001f83b4`)

The new "fix test" commit is a clean test infrastructure refactor. Replaces hardcoded `time.Sleep` + settle patterns with retry-based `faucetAndSettle` and `waitForSettle` helpers. No protocol-critical changes.

**Minor nits (non-blocking):**

1. **`utils_test.go:873`** — Missing space before `{`: `amount float64){` → `amount float64) {`. Run `gofmt`.
2. **`utils_test.go:883`** — Unnecessary bare `return` at end of `faucetAndSettle`.
3. **`vhtlc_test.go:841`** — `faucetAndSettle` already calls `strings.TrimSpace(address)` internally, but this call site double-trims with `strings.TrimSpace(boarding[0])`. Other call sites don't. Pick one pattern.
4. **`main_test.go:106-110`** — The closure `func(ctx context.Context) error { _, err := f.Settle(ctx, &pb.SettleRequest{}); if err != nil { return err }; return nil }` simplifies to `_, err := f.Settle(...); return err`.

All test-only, all non-blocking. LGTM.

**Comment by Kukks** [comment]
@sekulicd  can you handle this? Needs to make sure no docker image from boltz is cached

**Review by ghost — CHANGES_REQUESTED** [review]
## 🤖 Reviewed by Arkana (incremental — new commit `471fd03`)

This "debug ci" commit has two changes. One is fine, one must not be merged.

### 🔴 `spew.Dump` left in production code

**`pkg/swap/chainswap_btc_ark_handler.go:50`** — `spew.Dump(update)` added to `HandleLockupFailed`. This is a debug statement that:
1. Dumps the full `SwapUpdate` struct (including potentially sensitive swap details) to stdout in production.
2. Adds a new dependency (`github.com/davecgh/go-spew/spew`) to a non-test package.
3. Bypasses the existing structured logging (`logrus`) already used in this file.

**Action:** Remove before merge. If you need to log the update on failure, use `log.WithField("update", update).Debug(...)` which respects log-level configuration.

### ✅ CI workflow — docker logs on failure

`.github/workflows/ci.integration.yaml` — Adding a failure-only step to dump docker logs is a clean debugging aid. No issues.

---

**Verdict:** Request changes — the `spew.Dump` must be removed from production code before merge.

**Review by ghost — APPROVED** [review]
## 🤖 Reviewed by Arkana (incremental — new commit `6fbfe5d9`)

Previous blocker resolved: `spew.Dump` and the `spew` import are removed from `pkg/swap/chainswap_btc_ark_handler.go`. Production code is clean.

No further issues. The overall PR is test-only + CI diagnostics — no protocol-critical production changes.

**LGTM** ✅

**Comment by sekulicd** [comment]
> @sekulicd can you handle this? Needs to make sure no docker image from boltz is cached

We already pulling images with make build-test-env which is part of CI
```
pull-test-env:
	@echo "Updating test env images..."
	@docker compose -f test.docker-compose.yml pull

build-test-env: pull-test-env
	@echo "Building test environment..."
	@docker compose -f test.docker-compose.yml build --no-cache
```
Test added by @michael1011 is proving that fix he added in Boltz is working, we dont need to rush with merging this cause there is flaky test related to chain swap:
```
=== RUN   TestChainSwapBTCtoARKWithQuote
    chainswap_test.go:106: Balance: 99837
    chainswap_test.go:108: Creating BTC→ARK chain swap...
    chainswap_test.go:116: Created chain swap: d4LhrWc2Ck11
    chainswap_test.go:122: Swap d4LhrWc2Ck11 error: failed to get quote: GET http://boltz:9001/v2/swap/chain/d4LhrWc2Ck11/quote: HTTP 400: {"error":"lockup transaction was not rejected because of the amount"}
    chainswap_test.go:122: 
        	Error Trace:	/home/runner/work/fulmine/fulmine/internal/test/e2e/chainswap_test.go:556
        	            				/home/runner/work/fulmine/fulmine/internal/test/e2e/chainswap_test.go:122
        	Error:      	Not equal: 
        	            	expected: "claimed"
        	            	actual  : "failed"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-claimed
        	            	+failed
        	Test:       	TestChainSwapBTCtoARKWithQuote
```
```
I added logs in CI in case it fails and here are logs from fulmine:
ulmine     | 2026-05-01T09:38:11.008983024Z time="2026-05-01T09:38:11Z" level=info msg="Created BTC→ARK chain swap HUmWF8b6ztRb with Boltz"
fulmine     | 2026-05-01T09:38:11.009039499Z time="2026-05-01T09:38:11Z" level=info msg="Please send 3000 sats to: bcrt1p903zhxy8v59yur5wzffzj2yd3avmuxs3w8fzfygexlpydrjlxzgquu5h52"
fulmine     | 2026-05-01T09:38:11.010830625Z time="2026-05-01T09:38:11Z" level=debug msg="Cached swap response for swap HUmWF8b6ztRb (used during active monitoring)"
fulmine     | 2026-05-01T09:38:11.010851684Z time="2026-05-01T09:38:11Z" level=info msg="Created chain swap HUmWF8b6ztRb: BTC → Ark, lockup address: bcrt1p903zhxy8v59yur5wzffzj2yd3avmuxs3w8fzfygexlpydrjlxzgquu5h52"
fulmine     | 2026-05-01T09:38:11.010856102Z time="2026-05-01T09:38:11Z" level=debug msg="method=/fulmine.v1.Service/CreateChainSwap duration=29ms request={\"amount\":3000,\"direction\":2}"
fulmine     | 2026-05-01T09:38:11.010944264Z time="2026-05-01T09:38:11Z" level=info msg="Starting WebSocket monitoring for chain swap HUmWF8b6ztRb"
fulmine     | 2026-05-01T09:38:11.013284299Z time="2026-05-01T09:38:11Z" level=info msg="Chain swap HUmWF8b6ztRb status update: 0 (raw: swap.created)"
fulmine     | 2026-05-01T09:38:11.013298886Z time="2026-05-01T09:38:11Z" level=info msg="Swap HUmWF8b6ztRb created, waiting for user to lock BTC"
fulmine     | 2026-05-01T09:38:11.047111032Z time="2026-05-01T09:38:11Z" level=debug msg="method=/fulmine.v1.Service/ListChainSwaps duration=0ms request={\"swap_ids\":[\"HUmWF8b6ztRb\"]}"
fulmine     | 2026-05-01T09:38:11.062976342Z time="2026-05-01T09:38:11Z" level=info msg="Providing cooperative signature for Boltz to claim BTC lockup for swap 5hqA3LwNFqC4"
fulmine     | 2026-05-01T09:38:11.082808408Z time="2026-05-01T09:38:11Z" level=info msg="Chain swap HUmWF8b6ztRb status update: 13 (raw: transaction.lockupFailed)"
fulmine     | 2026-05-01T09:38:11.082823346Z time="2026-05-01T09:38:11Z" level=warning msg="User lockup failed for swap HUmWF8b6ztRb (amount mismatch), fetching quote"
fulmine     | 2026-05-01T09:38:11.106255517Z time="2026-05-01T09:38:11Z" level=error msg="Event handler error for swap HUmWF8b6ztRb" error="failed to get quote: GET http://boltz:9001/v2/swap/chain/HUmWF8b6ztRb/quote: HTTP 400: {\"error\":\"lockup transaction was not rejected because of the amount\"}"
fulmine     | 2026-05-01T09:38:11.125183965Z time="2026-05-01T09:38:11Z" level=info msg="Successfully provided cooperative signature for Boltz to claim BTC"
fulmine     | 2026-05-01T09:38:11.125196408Z time="2026-05-01T09:38:11Z" level=info msg="Successfully provided cooperative signature for Boltz BTC claim"
fulmine     | 2026-05-01T09:38:11.550390913Z time="2026-05-01T09:38:11Z" level=debug msg="method=/fulmine.v1.Service/ListChainSwaps duration=0ms request={\"swap_ids\":[\"HUmWF8b6ztRb\"]}"
```

boltz logs:

```
boltz          | 2026-05-01T09:38:10.985370405Z 01/05/2026 09:38:10:985 verbose: Creating new Taproot Chain Swap from BTC to ARK: HUmWF8b6ztRb
boltz          | 2026-05-01T09:38:11.006113035Z 01/05/2026 09:38:11:005 debug: Swap HUmWF8b6ztRb update: {
boltz          | 2026-05-01T09:38:11.006133854Z   "status": "swap.created"
boltz          | 2026-05-01T09:38:11.006137049Z }
boltz          | 2026-05-01T09:38:11.008174841Z 01/05/2026 09:38:11:008 verbose: Created Chain Swap with id: HUmWF8b6ztRb
boltz          | 2026-05-01T09:38:11.039239788Z 2026-05-01T09:38:11.038993Z  INFO boltzr::swap::tx_check: boltzr/src/swap/tx_check.rs:86: Found unconfirmed output for Swap HUmWF8b6ztRb in BTC transaction: 3bddc77f22037f35b85929205a1ca91ca7095377daddb3f129f7a5d44ea9dc19
boltz          | 2026-05-01T09:38:11.052608923Z 01/05/2026 09:38:11:052 verbose: Found unconfirmed BTC lockup transaction for Chain Swap HUmWF8b6ztRb: 3bddc77f22037f35b85929205a1ca91ca7095377daddb3f129f7a5d44ea9dc19:0
boltz          | 2026-05-01T09:38:11.055489630Z 2026-05-01T09:38:11.051170Z  INFO boltzr::swap::tx_check: boltzr/src/swap/tx_check.rs:86: Found confirmed output for Swap HUmWF8b6ztRb in BTC transaction: 3bddc77f22037f35b85929205a1ca91ca7095377daddb3f129f7a5d44ea9dc19
boltz          | 2026-05-01T09:38:11.055504958Z 2026-05-01T09:38:11.054975Z DEBUG boltzr::swap::utxo_nursery: boltzr/src/swap/utxo_nursery.rs:219: Adding BTC block 163: 413574459a10ca3858159c7ee1b197fbbe6f70145feb6bfac8eb21c8d2f6d746
boltz          | 2026-05-01T09:38:11.055735476Z 01/05/2026 09:38:11:055 debug: Got BTC block 163 (413574459a10ca3858159c7ee1b197fbbe6f70145feb6bfac8eb21c8d2f6d746) from sidecar
boltz          | 2026-05-01T09:38:11.076217239Z 01/05/2026 09:38:11:076 warn: Lockup of Chain Swap HUmWF8b6ztRb failed: locked 15500 is more than expected 3000
boltz          | 2026-05-01T09:38:11.079882405Z 01/05/2026 09:38:11:079 debug: Swap HUmWF8b6ztRb update: {
boltz          | 2026-05-01T09:38:11.079896671Z   "status": "transaction.lockupFailed",
boltz          | 2026-05-01T09:38:11.079899857Z   "failureReason": "locked 15500 is more than expected 3000",
boltz          | 2026-05-01T09:38:11.079902252Z   "failureDetails": {
boltz          | 2026-05-01T09:38:11.079904456Z     "actual": 15500,
boltz          | 2026-05-01T09:38:11.079906680Z     "expected": 3000
boltz          | 2026-05-01T09:38:11.079908984Z   }
boltz          | 2026-05-01T09:38:11.079910978Z }
boltz          | 2026-05-01T09:38:11.086698093Z 01/05/2026 09:38:11:086 verbose: Found confirmed BTC lockup transaction for Chain Swap HUmWF8b6ztRb: 3bddc77f22037f35b85929205a1ca91ca7095377daddb3f129f7a5d44ea9dc19:0
boltz          | 2026-05-01T09:38:11.101948512Z 01/05/2026 09:38:11:101 warn: Lockup of Chain Swap HUmWF8b6ztRb failed: locked 15500 is more than expected 3000
boltz          | 2026-05-01T09:38:11.103709952Z 01/05/2026 09:38:11:103 warn: Request GET /v2/swap/chain/HUmWF8b6ztRb/quote failed: {"error":"lockup transaction was not rejected because of the amount"}
boltz          | 2026-05-01T09:38:11.108163400Z 01/05/2026 09:38:11:108 debug: Swap HUmWF8b6ztRb update: {
boltz          | 2026-05-01T09:38:11.108177055Z   "status": "transaction.lockupFailed",
boltz          | 2026-05-01T09:38:11.108183497Z   "failureReason": "locked 15500 is more than expected 3000",
boltz          | 2026-05-01T09:38:11.108186322Z   "failureDetails": {
boltz          | 2026-05-01T09:38:11.108188536Z     "actual": 15500,
boltz          | 2026-05-01T09:38:11.108190820Z     "expected": 3000
boltz          | 2026-05-01T09:38:11.108193035Z   }
boltz          | 2026-05-01T09:38:11.108195259Z }
```

Boltz’s API getQuote is throwing error sometimes and @michael1011 is checking how we can improve.
In localhost where communication between services is faster this test usually passes but in GHA which is slower we have intermittent failure, could be some race issue which we need to explore anyways.
@Kukks
